pkg/virt-launcher/virtwrap/network/podinterface.go

```diff
@@ -254,11 +255,10 @@ func (l *podNICImpl) PlugPhase2(vmi *v1.VirtualMachineInstance, iface *v1.Interf
 		return err
 	}
 
-	isExist, err := bindMechanism.loadCachedInterface()
-	if err != nil {
+	if err := bindMechanism.loadCachedInterface(); err != nil {
 		log.Log.Reason(err).Critical("failed to load cached interface configuration")
 	}
-	if !isExist {
+	if !bindMechanism.hasCachedInterface() {
```

Comment:
I have to admit the even with this new `hasCachedInterface` I find this code to be very confusing.
What do you think about changing `loadCachedInterface` to be a global method that return the interface (pointer).

If the returned interface in not `nil` setting the relevant `BindMechanism` interface. If not, calling the `prepare`, `dicover, etc, methods.

Reply:
I am working on a follow up PR where I move all interface cache methods away from the bind mechanism interface, and they kind of do that (I move load/set CachedInterface to be methods of podNICImpl). 

If I follow your suggestion, what do I do w/ the `setCachedInterface` ? Also remove it from the bind mechanism interface, making it a function ? What do I do with slirp on that particular scenario ?

Reply:
Just add a special `if` For slirp. If it is slirp, not invoke the caching code.

Reply:
We better avoid a condition on a specific binding, it does not scale well and we will need to know this one when adding a new binding.

Reply:
> We better avoid a condition on a specific binding, it does not scale well and we will need to know this one when adding a new binding.

I agree; that is why interfaces have different implementations.

What if the `loadCachedInterface` remains in the interface (for now) and I update its signature to `loadCachedInterface() (*api.Interface, error)` ?

Reply:
We already have this `if` around the [setPodInterfaceCache](https://github.com/kubevirt/kubevirt/blob/a59568d288f64639f6f45e59975713942a14a786/pkg/virt-launcher/virtwrap/network/podinterface.go#L193).
If we want to make `PlugPhase1` generic. We can move it (most of it) to be a bind mechanism methods.
`Slirp` doesn't need 99% of the `PlugPhase1` code.

Reply:
> Slirp doesn't need 99% of the `PlugPhase1` code.

I agree; I'd even go as far as to say Slirp doesn't need anything of the phase1. In fact, imho, slirp is not a bind mechanism - it just needs to decorate the domxml. But that is besides the point; one of my long term goals for this refactoring is splitting the bind mechanism interface into 3: 
 - 1 configuring the networking infrastructure
 - 1 generating the domxml
 - 1 starting dhcp / whatever dynamic addressing mechanism we have.

But none of this has any impact on the decision at hand, nor this PR. At most, it just helps scoping this change.

> We already have this if around the setPodInterfaceCache.

The fact that we already have something weird in the code (the if around [setPodInterfaceCache](https://github.com/kubevirt/kubevirt/blob/a59568d288f64639f6f45e59975713942a14a786/pkg/virt-launcher/virtwrap/network/podinterface.go#L193)) shouldn't be a reason for us to make it more weird / harder to reason about. We should be making it easier - e.g. removing that if iface.slirp from there.

I do think what I'm proposing here is more readable that what we currently have, but am trying to adjust your comment that it is not readable enough into account. 

What is wrong w/ my last suggestion ? (What if the loadCachedInterface remains in the interface (for now) and I update its signature to `loadCachedInterface() (*api.Interface, error)` ?

By doing it, we hide the fact that slirp is a special scenario on the implementation of its methods, where, imho, it should be. This allows for the `PlugPhase1` method to be genetic, instead of concerning itself w/ which binding type it is using.

Reply:
I just want to show how this will look like with the aforementioned implementation.

I think it is more complex, and will still require changes for the unit tests to pass.

While this is the direction that I have in mind, I hope I can convince you using the `hasCachedInterface` is a solid compromise between the mess we currently have and what we want to have in the future.